### PR TITLE
Validate ImageChangeTriggers of DockerStrategy BuildConfigs

### LINKED ISSUE #4447: BuildConfig validation allows invalid strategies, causes panics in build controller
Extracted from https://github.com/openshift/origin/pull/4416#issuecomment-136673792

This file creates a BuildConfig object that passes validation, but causes panics when the BuildController tries to instantiate the build.

eg. [kill.yaml](https://gist.githubusercontent.com/nak3/6387633d1e833929cc2d/raw/3ac14d7f4a9569fba7c3ee01daa8ba3d823712e5/killer.yaml)
~~~
oc create -f https://gist.githubusercontent.com/nak3/6387633d1e833929cc2d/raw/3ac14d7f4a9569fba7c3ee01daa8ba3d823712e5/killer.yaml
~~~

- origin.log
~~~
I0901 19:44:25.260717    7279 config_controller.go:38] Running build for BuildConfig default/hello-c
I0901 19:44:25.269988    7279 generator.go:167] Generating Build from BuildConfig: default/hello-c, LastVersion: 0
E0901 19:44:25.271091    7279 util.go:72] Recovered from panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/tmp/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/util.go:66
/tmp/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/util.go:54
/usr/lib/golang/src/runtime/asm_amd64.s:402
/usr/lib/golang/src/runtime/panic.go:387
/usr/lib/golang/src/runtime/panic.go:42
/usr/lib/golang/src/runtime/sigpanic_unix.go:26
/tmp/origin/_output/local/go/src/github.com/openshift/origin/pkg/build/generator/generator.go:237
/tmp/origin/_output/local/go/src/github.com/openshift/origin/pkg/build/generator/generator.go:177
/tmp/origin/_output/local/go/src/github.com/openshift/origin/pkg/build/registry/instantiate/generator/generator.go:35
~~~

Looks like conversion is setting dockerStrategy to an empty struct, so it passes the nil check, but is missing required parameters used during Instantiate.

Validation should prevent that object from getting persisted

### PR DESCRIPTION
**Pull request by rhcarvalho**
Fixes #4447 

@bparees PTAL

### CODE REVIEW

File: pkg/build/api/validation/validation.go [hunk 1]
```diff
@@ -76,6 +76,17 @@ func ValidateBuildConfig(config *buildapi.BuildConfig) fielderrors.ValidationErr
 	}
 
 	allErrs = append(allErrs, validateBuildSpec(&config.Spec.BuildSpec).Prefix("spec")...)
+
+	// validate ImageChangeTriggers of DockerStrategy builds
```
Comment (rhcarvalho) [inline pkg/build/api/validation/validation.go]: @bparees does it make sense here to also try to validate the `ObjectReference`(s), `DockerStrategy.From` and `ImageChange.From`, when they're not `nil`?

File: pkg/build/api/validation/validation_test.go [hunk 2]
```diff
@@ -144,7 +143,10 @@ func TestValidateBuildUpdate(t *testing.T) {
 	}
 }
 
-func TestBuildConfigValidationSuccess(t *testing.T) {
```
Comment (rhcarvalho) [inline pkg/build/api/validation/validation_test.go]: @bparees it turned out that we had a test that was wrong. Would you like us to keep another `TestBuildConfigValidationSuccess` that is actually a case of valid BC?

**Comment by rhcarvalho** [comment]
[test]

File: pkg/build/api/validation/validation_test.go (hunk 2 above)
Comment (bparees) [inline pkg/build/api/validation/validation_test.go]: no we can probably live without that.. if we manage to break successful validation i think we'll find out pretty fast :)

File: pkg/build/api/validation/validation.go (hunk 1 above)
Comment (bparees) [inline pkg/build/api/validation/validation.go]: validateTrigger validates ImageChange.From
validate[Source/Custom/Docker]Strategy validates the Strategy.From

**Comment by bparees** [comment]
lgtm. [merge]

**Pull request merged**